libzecale/circuits/fields/fp6_3over2_gadgets.tcc

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef ZECALE_CIRCUITS_FIELDS_FP6_3OVER2_GADGETS_TCC__
+#define ZECALE_CIRCUITS_FIELDS_FP6_3OVER2_GADGETS_TCC__
+
+#include "libzecale/circuits/fields/fp6_3over2_gadgets.hpp"
+
+namespace libzecale
+{
+
+// Fp6_3over2_variable methods
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb, const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(pb, FMT(annotation_prefix, " c0"))
+    , _c1(pb, FMT(annotation_prefix, " c1"))
+    , _c2(pb, FMT(annotation_prefix, " c2"))
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp6_3over2_variable<Fp6T> &v,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(v._c0)
+    , _c1(v._c1)
+    , _c2(v._c2)
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp6T &v,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(pb, v.c0, FMT(annotation_prefix, " c0"))
+    , _c1(pb, v.c1, FMT(annotation_prefix, " c1"))
+    , _c2(pb, v.c2, FMT(annotation_prefix, " c2"))
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T>::Fp6_3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const libsnark::Fp2_variable<Fp2T> &c0,
+    const libsnark::Fp2_variable<Fp2T> &c1,
+    const libsnark::Fp2_variable<Fp2T> &c2,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix), _c0(c0), _c1(c1), _c2(c2)
+{
+}
+
+template<typename Fp6T>
+Fp6_3over2_variable<Fp6T> Fp6_3over2_variable<Fp6T>::operator*(const FieldT &a)
+{
+    return Fp6_3over2_variable<Fp6T>(
+        this->pb,
+        _c0 * a,
+        _c1 * a,
+        _c2 * a,
+        FMT(this->annotation_prefix, " * Fr"));
```

Comment:
This gadget is the operator* of an **extension field** built over the scalar field (commonly denoted Fr) of BW6. May be worth being a bit more precise in the name of the annotation here - to have clean logs and facilitate the debug of the circuit

Reply:
In this case, it is a simple linear combination representing a multiple by a constant scalar.  I've switched to " *Fp" for consistency with the class name.